drivers/led/lp50xx.c

```diff
@@ -0,0 +1,288 @@
+
+struct lp50xx_config {
+	struct i2c_dt_spec bus;
+	uint8_t num_leds;
+	bool log_scale_en;
+	bool max_curr_opt;
+	const struct led_info *leds_info;
+	enum lp50xx_chips chip;
+};
+
+struct lp50xx_data {
+	uint8_t *chan_buf;
+};
+
+static const struct led_info *
+lp50xx_led_to_info(const struct lp50xx_config *config, uint32_t led)
+{
+	for (uint8_t i = 0; i < config->num_leds; i++) {
+		if (config->leds_info[i].index == led) {
+			return &config->leds_info[i];
+		}
+	}
+	return NULL;
+}
+
+static int lp50xx_get_info(const struct device *dev, uint32_t led,
+			const struct led_info **info)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	*info = led_info;
+
+	return 0;
+}
+
+static int lp50xx_set_brightness(const struct device *dev,
+				 uint32_t led, uint8_t value)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+	uint8_t buf[2];
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (value > 100) {
+		LOG_ERR("Exceeded brightness max value");
+		return -EINVAL;
+	}
+
+	buf[0] = LP50XX_LED_BRIGHTNESS_BASE(config->chip) + led_info->index;
+	buf[1] = (value * 0xff) / 100;
+
+	return i2c_write_dt(&config->bus, buf, sizeof(buf));
+}
+
+static int lp50xx_on(const struct device *dev, uint32_t led)
+{
+	return lp50xx_set_brightness(dev, led, 100);
+}
+
+static int lp50xx_off(const struct device *dev, uint32_t led)
+{
+	return lp50xx_set_brightness(dev, led, 0);
+}
+
+static int lp50xx_set_color(const struct device *dev, uint32_t led,
+			    uint8_t num_colors, const uint8_t *color)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+	uint8_t buf[4];
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (num_colors != led_info->num_colors) {
+		LOG_ERR("Provided number of colors is not valid");
+		return -EINVAL;
+	}
+
+	buf[0] = LP50XX_OUT_COLOR_BASE(config->chip) + 3 * led_info->index;
+	buf[1] = color[0];
+	buf[2] = color[1];
+	buf[3] = color[2];
+
+	return i2c_write_dt(&config->bus, buf, sizeof(buf));
+}
+
+static int lp50xx_write_channels(const struct device *dev,
+				 uint32_t start_channel,
+				 uint32_t num_channels, const uint8_t *buf)
+{
+	const struct lp50xx_config *config = dev->config;
+	struct lp50xx_data *data = dev->data;
+
+	if (start_channel >= LP50XX_NUM_CHANNELS(config->chip) ||
+	    start_channel + num_channels > LP50XX_NUM_CHANNELS(config->chip)) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Unfortunately this controller doesn't support commands split into
+	 * two I2C messages.
+	 */
+	data->chan_buf[0] = LP50XX_CHANNEL_BASE(config->chip) + start_channel;
+	memcpy(data->chan_buf + 1, buf, num_channels);
+
+	return i2c_write_dt(&config->bus, data->chan_buf, num_channels + 1);
+}
+
+enum lp50xx_chips lp50xx_get_chip_id(const struct device *dev)
```

Comment:
NAK. Sharing a custom function between the driver and samples is entirely wrong and violates the LED API.

I think the issue you are trying to solve with this is, supporting a single sample for different LED chips. You could use separate samples also.